test/ports_test.dart

```diff
@@ -22,259 +22,335 @@ void main() {
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       fail('unreachable');
     }, onError: (e, s) {
       expect(e, 92);
     });
   });
 
   test('TimeoutCallbackSLow', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100,
         onTimeout: () => Future.delayed(_ms * 500, () => 87));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       expect(v, 87);
     });
   });
 
   test('TimeoutCallbackThrowsSlow', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100,
         onTimeout: () => Future.delayed(_ms * 500, () => throw 87));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       fail('unreachable');
     }, onError: (e, s) {
       expect(e, 87);
     });
   });
 
   test('TimeoutFirst', () {
-    var completer = Completer.sync();
-    var p =
+    final completer = Completer.sync();
+    final p =
         singleCompletePort(completer, timeout: _ms * 100, onTimeout: () => 37);
     Timer(_ms * 500, () => p.send(42));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       expect(v, 37);
     });
   });
+
+  test('TimeoutFirst with valid null', () {
+    final completer = Completer<int?>.sync();
+    final p = singleCompletePort(completer,
+        timeout: _ms * 100, onTimeout: () => null);
+    Timer(_ms * 500, () => p.send(42));
+    return expectLater(completer.future, completion(null));
+  });
+
+  test('TimeoutFirst with invalid null', () {
+    final completer = Completer<int>.sync();
+
+    /// example of incomplete generic parameters promotion.
+    /// same code with [singleCompletePort<int, dynamic>] is a compile time error
+    final p = singleCompletePort(
+      completer,
+      timeout: _ms * 100,
+      onTimeout: () => null,
+    );
+    Timer(_ms * 500, () => p.send(42));
+    return expectLater(completer.future, throwsA(isA<TypeError>()));
+  });
 }
 
 void testSingleResponseFuture() {
   test('FutureValue', () {
     return singleResponseFuture((SendPort p) {
       p.send(42);
-    }).then((v) {
+    }).then<Null>((v) {
+      expect(v, 42);
+    });
+  });
+
+  test('FutureValueWithoutTimeout', () {
+    return singleResponseFutureWithoutTimeout<int>((SendPort p) {
+      p.send(42);
+    }).then<Null>((v) {
       expect(v, 42);
     });
   });
 
+  test('FutureValueWithoutTimeout valid null', () {
+    return singleResponseFutureWithoutTimeout<int?>((SendPort p) {
+      p.send(null);
+    }).then<Null>((v) {
+      expect(v, null);
+    });
+  });
+
+  test('FutureValueWithoutTimeout invalid null', () {
+    return expectLater(singleResponseFutureWithoutTimeout<int>((SendPort p) {
+      p.send(null);
+    }), throwsA(isA<TypeError>()));
+  });
```

Comment:
this is impossible to check statically since `SendPort.send` expects `Object?`

Reply:
Yes, ports are inherently untyped. Primarily because they are able to send values between isolates that don't necessarily share types.

We have to get the value to the other side of the port connection in order to check it against the original type argument.